### PR TITLE
Schema updates

### PR DESCRIPTION
**Pull request by glatard**
Schema updates to address #69, #72 and #79, with corresponding examples and documentation.

### CODE REVIEW

**Comment by glatard** [comment]
@kofalt may have an opinion on this too.

File: schema/descriptor.schema.json [hunk 1]
```diff
@@ -85,7 +91,7 @@
             "id": "http://github.com/boutiques/boutiques-schema/schema-version",
             "type": "string",
             "description": "Version of the schema used.",
-            "enum": ["0.2", "0.3"]
+            "enum": ["0.4"]
```
Comment (kofalt) [inline schema/descriptor.schema.json]: So this raises an interesting question - if a system wants backwards compatibility with multiple boutique versions, would the approach have to be first, take each released json schema to see which (if any) match, and go from there?

My intuition says "yes"; curious what you think :smile:

**Comment by kofalt** [comment]
Changes look positive, I like it. 

I have a comment on #69 that I should have written on that ticket & is my fault for not writing earlier. Seeing how the invocation schema is a 100% generated field, it should be A) always possible to generate from a boutique doc, B) possible to have an error in what the author wrote!

For these reasons I'm tempted to think of the invocation schema as a layer above the alg specification. This is our current approach in the Exchange for both [boutiques](https://github.com/flywheel-io/exchange/blob/master/manifests/alpine-sha384-159fa0a4649f1eb5c9603fdacbcf4e2ff1676dce8b163b7925a0b83613b8f49e38b4be708df063a15b28df3db3988c71.json) and [gears](https://github.com/flywheel-io/exchange/blob/master/manifests/scitran/scitran-dcm2niix-sha384-e61739c42fcbd5370f15c380860efd5b65a5961b81fb6b110305dc5195c5c09c668770cd0c897731ebe5ef228cefc1e2.json); we generate the invocation schema when the Exchange receives your document and stores it - separately - for convenience and archival purposes.

This means that the only stored invocation schema is generated by a submission process that will do it the same time every way, and removes the question of "what do I do when a boutique has an invalid or non-standard invocation schema".

TL;DR I would support **not** having an invocation schema on the boutique itself and it's the system's responsibility to either generate it on-the-fly or at the point of archival, to remove possible errors & confusion. Thoughts?

File: schema/descriptor.schema.json (hunk 1 above)
Comment (kofalt) [inline schema/descriptor.schema.json]: Sounds good to me :+1:

**Comment by glatard** [comment]
I understand your point but, although we haven't played a lot with invocation schemas so far, I see a few issues:
* Generating the invocation schema requires a specific tool, currently the invocationSchemaHandler.py. It is much easier for a platform to validate inputs using straight JSON schema validation (which can be done in any language) than to have to integrate a Python script somewhere in the process. Ideally, invocation schemas should be generated asap, ideally every time there is an edit in the manifest (that's why it's so cool to have a repo that can trigger such operations). If platforms have to do it, it will only add complexity.
* Keeping the schema and manifest separated will create issues when both documents are inconsistent (what if you're testing 2 versions of the same manifest with incompatible schemas, etc).

That's why I think including the invocation schema in the manifest should be the norm. However, if for some reasons a platform wants to ignore this property and re-generate the invocation schemas to store them separately (or extract the schemas and store them in a different document), it is always possible.

What do you think?

**Comment by kofalt** [comment]
My main concern is that an author adds some option (or something) and forgets to update his invocation - when I'm processing that boutique, do I honor their (invalid) schema? Generate my own? Why even have that possibility? It's the difference between the system being invalid: go debug it, or the author being invalid: send an email :/

Not needing python to run a boutique is a good feature though, it's one of the reasons the Exchange generates the invocation & stores it. If this lands, maybe we should update the Exchange will reject boutiques that are submitted with a schema that differs from that spec version?

I would be fine going ahead with this as long as it's clear that the invocation should always match what we expect to generate, and that consumers have the option of using their own variant of the invocation, which is generally what I would do :smile:

**Comment by glatard** [comment]
I agree that inconsistencies between the manifest and the invocation schema could still be a problem. If I received a manifest with an invocation schema from a developer (human), I would definitely double-check it.

Given that the handling of invocation schemas is still a bit nascent, I would suggest that we keep this property for now, and that

**Comment by kofalt** [comment]
> for now, and that 

I think your comment cut off :) But yeah, I'm cool with that.
